Add max_energy_range_uj to deal with RAPL wraparound

mp15 opened the pull request:
When we detect a negative delta from RAPL energy usage then assume we have wrapped around and add max_energy_range_uj to it. According to Intel docs this should happen about every 60 secs under a heavy load. Fixes #322

mp15 commented:
We probably should write a test to verify the delta functionality with mock rollover readings but this should be better than returning 0.

benoit-cty commented:
Thank you very much for this investigation !

If we understand well:
- RAPL write energy to a counter `energy_uj`.
- At some point the counter `energy_uj` goes above the constant `max_energy_range_uj` and fall back to zero.
- RAPL continue to write energy to `energy_uj`.

So if it goes to zero while codecarbon is running, we have the negative value problem.

With your patch codecarbon will add `max_energy_range_uj` to `last_energy` in case of negative energy to avoid the problem.

May I suggest to refactor your code like this ?
```py
        if self.last_energy > energy:
            logger.debug(
                f"In RAPLFile : Current energy value ({energy}) is lower than previous value ({self.last_energy}). Assuming wrap-around! Source file : {self.path}"
            )
            energy = energy + self.max_energy_reading
          self.power = self.power.from_energies_and_delay(
              energy, self.last_energy, duration
          )
          self.energy_delta = energy - self.last_energy
          self.last_energy = energy
          return
```

mp15 commented:
I've followed your refactor but I have added a new_last_energy variable to it so that it works correctly the next time we call delta(duration). I assume we are accumulating the delta values across calls?
I've also modified the test to compare the results and use almostEqual (as Energy is held as a float internally and non-zero floats are fun to compare).

SaboniAmine commented on codecarbon/core/cpu.py:
```diff
@@ -201,11 +201,16 @@ def _fetch_rapl_files(self):
                     name = f"Processor Energy Delta_{i}(kWh)"
                     i += 1
                 rapl_file = os.path.join(self._lin_rapl_dir, file, "energy_uj")
+                rapl_file_max = os.path.join(
```
Would you please add a brief comment to describe why this additional rapl file is needed ?

SaboniAmine commented on tests/test_energy.py:
```diff
@@ -12,20 +12,28 @@ def setUp(self):
+        with open(self.max_path, "w") as f:
+            f.write("105")
 
     def test_negative_delta_writes_no_delta(self):
```
Would you please update this test's name (and the comment below, line 36) ?

SaboniAmine reviewed (approved):
Great work, thanks!
This corrects the previous patch we made on this and we have learnt something. 
Thanks for your contribution!

benoit-cty commented:
Thanks, you're right for `new_last_energy` :+1:
